### ISSUE TITLE
$TERM in :terminal is xterm-256color, but :term's escape codes aren't xterm-256 compatible

### ISSUE THREAD

**Issue by cmcaine** [neovim/neovim#7047]
- `nvim --version`:

```
NVIM v0.2.0
Build type: Release
Compilation: /usr/bin/cc -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -Wconversion -DNVIM_MSGPACK_HAS_FLOAT32 -O2 -DNDEBUG -DDISABLE_LOG -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -I/build/neovim/src/build/config -I/build/neovim/src/neovim-0.2.0/src -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/build/neovim/src/build/src/nvim/auto -I/build/neovim/src/build/include
Compiled by builduser

Optional features included (+) or not (-): +acl   +iconv    +jemalloc +tui      
For differences from Vim, see :help vim-differences

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"
```

**tl;dr** :terminal should report TERM=vte-256color. nvim's own documentation suggests this for other libVTE terminals: https://neovim.io/doc/user/term.html#TERM

**more detail**

libVTE isn't fully compatible with xterm-256color's escape codes but most libvte terminals still report their TERM as xterm-256color. This is especially annoying when combined with ssh as other indicator environment variables will be stripped.

The example I encounter most often: most terminals (including VTE) accept ^H as backsapce and ^? as ctrl-backspace. In xterm it's the other way around.

Another: libVTE supports true colour, xterm doesn't. That's maybe not relevant now (I don't think :terminal passes through true colour sequences?), but it could be in the future.

Anyway, nvim should set a different TERM and export an appropriate VTE_VERSION (the non-standard environment variable commonly used for testing if a terminal is really VTE rather than xterm).

### Steps to reproduce using `nvim -u NORC`

```
$ nvim -u NORC
:term cat
<C-v><BS>
# result:
^?

$ xterm
# in xterm
$ cat
<C-v><BS>
# result
^H
```

### Expected behaviour

TERM=vte-256color
VTE_VERSION=\<some number\>

**Comment by cmcaine** [comment]
There's already a terminfo file for vte terminals. Edited OP.

**tl;dr** :terminal should report TERM=vte-256color. nvim's own documentation suggests this for other libVTE terminals: https://neovim.io/doc/user/term.html#TERM

**Comment by justinmk** [comment]
kbs is a termios setting that we intentionally change, to distinguish it from delete.

**Comment by cmcaine** [comment]
BS, del and ctrl+BS all give different codes in my nvim terminal, or am I missing the point?

**Comment by justinmk** [comment]
Regardless, why would that make VTE a better fit? libvterm tries to be xterm compatible, not VTE.

**Comment by cmcaine** [comment]
I conflated libvte and libvterm. Sorry.

IMO, nvim should still set something other than xterm and ship with appropriate terminfo. Something like the below.

```
TERM=nvim
TERM=nvim-256color
TERM=vterm
```

**Comment by justinmk** [comment]
That will break all kinds of applications. The goal is to be fully xterm-compliant. Anything else is a bug, not a reason to declare a different TERM.

**Comment by cmcaine** [comment]
xterm supports all kinds of arcane stuff. I don't think it is wise or necessary to seek full compatibility.

Any sensibly written software uses terminfo to detect capabilities or sticks to more or less standard functionality. Changing TERM would not break either of those, provided terminfo files are distributed (which can be done in user dotfiles if it's not shipped by your distro, so even if you ssh from :term, you should be fine).

Some badly written software that only tests a few well-known TERM values will break. I almost never encounter code like that. YMMV.

From a developer PoV, terminals pretending to be xterm is the source of a whole variety of bugs and ugly hacks because these terminals are never fully xterm complaint (bugs and all) and those differences are occasionally important (Ctrl keycodes, 24bit colour support, all the complaints about gnome-terminal, etc).

The whole terminal ecosystem is a surprisingly well-performing mess, but within its constraints, I think the most correct course is to declare a different TERM.

If, however, you're sure you want to pretend to be xterm, can I open a PR on libvterm to make BS behaviour xterm-compatible?

**Comment by justinmk** [comment]
> Any sensibly written software uses terminfo to detect capabilities

Not really. `nvim` as an application tried that, it's garbage. terminfo sucks, linux/bsd distros can't agree on terminfo entries, and users don't know how to set or check their own terminfo files.

> or sticks to more or less standard functionality. 

That's what I meant by "xterm compliant". But really I just trust that @leonerd does the right thing.

> Changing TERM would not break either of those, provided terminfo files are distributed (which can be done in user dotfiles if it's not shipped by your distro, so even if you ssh from :term, you should be fine).

What is gained?

> If, however, you're sure you want to pretend to be xterm, can I open a PR on libvterm to make BS behaviour xterm-compatible?

kbs is a stty setting, a difference there does not constitute xterm incompatibility, AFAIK. Anyways, that's up to @leonerd

**Comment by justinmk** [comment]
@cmcaine Does https://github.com/neovim/neovim/pull/6992 help your issue?

**Comment by ZyX-I** [comment]
Note that if you `ssh` from `:term` you need terminfo *on the remote side*, not on your. “User dotfiles” is not always an option, and it *definitely* will harm user experience at first runs as terminfo will not be present. The request would’ve made sense if there already was not a bunch of terminals lying about who they are.

// BTW, when I was creating libvterm-based tests for powerline I added `st-256color` terminfo entry. Using any `xterm*` at the time made libvterm print something to stderr due to tmux output. Do not know whether something changed now.

**Comment by cmcaine** [comment]
@justinmk #6992 doesn't help: BS and C-BS give the same results as before in :term inside urxvt and gnome-terminal and strictly wrong results in xterm (^? for both BS and C-BS).

> What was gained?

libvterm and xterm would have different terminfo files because kbs is a terminfo attribute (though I think it should have the same value as stty erase). Other attributes would probably also differ. This would fix my immediate problem (I can make good-enough assumptions about C-BS based on the BS key).

It's also, like, totally good for the environment, man: terminfo is crap and outdated, but we can't build a replacement while terminals routinely lie in the one (non hacky) environment variable that reliably gets passed through ssh. A replacement could support all that non-terminfo covered stuff, like modifier keys, 24 bit colour, and bugs (xterm is really buggy for me).

I think nvim/vterm could get a terminfo entry + equivalents for the other libraries (if req) into the main distributions and switch the default TERM value in <5 years without much fuss.

This isn't my hill to die on, so I'm just going to fix it locally with some ugly hacks :)

@ZyX-I:

Copying terminfo from one machine to another is pretty easy. Though the commands are pretty obscure and undiscoverable (I got them from urxvt(7)), but maybe nvim could display a warning.

`infocmp $term | ssh $dest "mkdir -p .terminfo && cat >/tmp/ti && tic /tmp/ti"`

**Comment by ZyX-I** [comment]
@cmcaine I know how to run shell commands by ssh and that it passes stdout. The point is that you *need* to do this while with $TERMs present in ncurses package you don’t, and this action is easy to forget, especially if you are somewhere where you do not have your shell configuration (i.e. no shell functions which will run similar\* command for you), or using Neovim `:terminal` with ssh for the first time. And, BTW, Neovim on Windows (not WSL or cygwin) is going to have `:terminal`, but pretty much sure `infocmp` will not be available there.

\* Definitely not *your* one, it will write to /usr/share/terminfo given a chance and does not clean up temporary file or care about it possibly being already present.

**Comment by cmcaine** [comment]
Didn't mean to suggest that you don't know how to do this, just that it's not very difficult on most setups. I know that wee script is a bit bad, but I basically never run it, so it doesn't need to be good.

nvim could provide a safe and platform independent function if it was a big deal that the terminfo wasn't commonly available, but if the change is phased in over a longer time period I *imagine* it will be OK.

> easy to forget

I have something like this in my .zshrc. Obviously won't help if none of your shell configs are on the remote, but would catch some cases.

```
if ! infocmp &>/dev/null; then
  echo "WARN: terminfo for your term is not available! Run ssh-copy-terminfo" 1>&2
fi
```
